File: ios/Sources/KlinikFinanceFeature/FinanceScreen.swift
import SwiftUI
import KlinikAPI
import KlinikCore
import KlinikDesign

/**
 * The money (spec M11).
 *
 * Two questions at the top because they are the two a clinic asks: what is
 * still owed, and what came in this month. The ledger is underneath, filtered
 * by status, because working through it is a different job from checking on it.
 *
 * The client performs no arithmetic. Every figure on this screen is a string
 * the server produced from a decimal; a balance recomputed here would be a
 * second opinion, and the two would eventually differ in front of a patient.
 */
public struct FinanceScreen: View {
    @Environment(\.colorScheme) private var scheme

    private let model: FinanceModel
    private let openPatient: (String, String) -> Void

    @State private var state = FinanceState()
    @State private var paying: FinanceRecord?
    @State private var reversing: PaymentEntry?

    public init(model: FinanceModel, openPatient: @escaping (String, String) -> Void) {
        self.model = model
        self.openPatient = openPatient
    }

    public var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: Tokens.Spacing.xl) {
                ErrorBanner(message: state.error)

                currencyPicker

                switch state.phase {
                case .loading:
                    VStack(spacing: Tokens.Spacing.lg) {
                        SkeletonCard(lines: 3)
                        SkeletonCard(lines: 4)
                    }
                    .accessibilityElement()
                    .accessibilityLabel(L10n.string("common.loading"))

                case .notPermitted:
                    MessageState(icon: "lock", text: L10n.string("finance.notPermitted"))
                        .frame(minHeight: 280)

                case .failed(let message):
                    MessageState(
                        icon: Tokens.State.labCritical.iconName,
                        text: message,
                        retryTitle: L10n.string("common.retry")
                    ) {
                        await reload()
                    }

                case .loaded:
                    headline
                    ageing
                    ledger
                }
            }
            .padding(Tokens.Spacing.lg)
        }
        .background(Tokens.Palette.background.resolve(for: scheme))
        .navigationTitle(L10n.string("finance.title"))
        .refreshable { await reload() }
        .task { await reload() }
        .sheet(item: $reversing) { payment in
            ReversalSheet(payment: payment) { reason in
                let ok = await model.reverse(paymentId: payment.id, reason: reason)
                state = model.currentState()

                if ok { reversing = nil }
            }
        }
        .sheet(item: $paying) { record in
            PaymentSheet(record: record) { amount, method, reference in
                let ok = await model.pay(
                    recordId: record.id,
                    amount: amount,
                    method: method,
                    reference: reference
                )
                state = model.currentState()

                if ok { paying = nil }
            }
        }
    }

    private var currencyPicker: some View {
        Picker(L10n.string("finance.currency"), selection: currencyBinding) {
            ForEach(Currency.allCases, id: \.self) { currency in
                Text("\(currency.symbol) \(currency.rawValue)").tag(currency)
            }
        }
        .pickerStyle(.segmented)
        .accessibilityLabel(L10n.string("finance.currency"))
    }

    private var currencyBinding: Binding<Currency> {
        Binding(
            get: { state.currency },
            set: { currency in
                Task {
                    await model.choose(currency: currency)
                    state = model.currentState()
                }
            }
        )
    }

    // MARK: - The two questions

    @ViewBuilder
    private var headline: some View {
        HStack(spacing: Tokens.Spacing.md) {
            if let outstanding = state.outstanding {
                StatTile(
                    value: outstanding.outstanding.converted.display(state.currency),
                    label: L10n.string("finance.outstanding"),
                    tone: outstanding.outstanding.converted.isZero ? .success : .warning,
                    symbol: "hourglass"
                )
            }

            if let collections = state.collections {
                StatTile(
                    value: collections.net.converted.display(state.currency),
                    label: L10n.string("finance.collectedThisMonth"),
                    tone: .success,
                    symbol: "arrow.down.circle"
                )
            }
        }

        // A total assembled from three currencies where one had no rate for its
        // day is not a total. Said beside the figure, not under it.
        if let outstanding = state.outstanding, !outstanding.outstanding.complete {
            Card(tone: .warning) {
                Label(
                    L10n.string("finance.totals.incomplete"),
                    systemImage: "exclamationmark.circle"
                )
                .font(Tokens.Typography.captionRelative)
                .foregroundStyle(Tokens.Palette.warning.resolve(for: scheme))
                .fixedSize(horizontal: false, vertical: true)
            }
        }
    }

    @ViewBuilder
    private var ageing: some View {
        if let report = state.outstanding, !report.ageing.isEmpty {
            VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
                SectionHeader(
                    title: L10n.string("finance.ageing"),
                    subtitle: String(format: L10n.string("finance.recordCount"), report.recordCount)
                )

                Card {
                    VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
                        ForEach(report.ageing) { bucket in
                            HStack {
                                Text(L10n.string("finance.ageing.\(bucket.bucket)"))
                                    .font(Tokens.Typography.calloutRelative)
                                    .foregroundStyle(
                                        Tokens.Palette.textPrimary.resolve(for: scheme)
                                    )

                                Spacer(minLength: Tokens.Spacing.sm)

                                Text(bucket.totals.converted.display(state.currency))
                                    .font(Tokens.Typography.subheadingRelative)
                                    .foregroundStyle(
                                        FinanceScreen.ageingTone(bucket.bucket)
                                            .foreground.resolve(for: scheme)
                                    )

                                Badge("\(bucket.recordCount)")
                            }
                            .accessibilityElement(children: .combine)
                        }
                    }
                }
            }
        }
    }

    // MARK: - The ledger

    private var ledger: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
            SectionHeader(title: L10n.string("finance.records"))

            Picker(L10n.string("finance.status"), selection: statusBinding) {
                Text(L10n.string("finance.allStatuses")).tag(PaymentStatus?.none)

                ForEach(PaymentStatus.allCases, id: \.self) { status in
                    Text(status.localizedName).tag(PaymentStatus?.some(status))
                }
            }
            .pickerStyle(.menu)
            .frame(minHeight: Tokens.minimumTouchTarget)
            .accessibilityLabel(L10n.string("finance.status"))

            if state.records.isEmpty {
                Card {
                    Text(L10n.string("finance.noRecords"))
                        .font(Tokens.Typography.bodyRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                }
            } else {
                ForEach(state.records) { record in
                    card(record)
                }

                if state.hasMore {
                    Button(L10n.string("finance.loadMore")) {
                        Task {
                            await model.loadMore()
                            state = model.currentState()
                        }
                    }
                    .font(Tokens.Typography.calloutRelative)
                    .frame(maxWidth: .infinity, minHeight: Tokens.minimumTouchTarget)
                    .foregroundStyle(Tokens.Palette.accent.resolve(for: scheme))
                }
            }
        }
    }

    private var statusBinding: Binding<PaymentStatus?> {
        Binding(
            get: { state.status },
            set: { status in
                Task {
                    await model.choose(status: status)
                    state = model.currentState()
                }
            }
        )
    }

    private func card(_ record: FinanceRecord) -> some View {
        Card(tone: record.balance.isZero ? .neutral : .warning) {
            VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
                if let patient = record.patient {
                    Button {
                        openPatient(patient.id, patient.fullName)
                    } label: {
                        HStack(spacing: Tokens.Spacing.md) {
                            InitialsAvatar(name: patient.fullName, diameter: 36)

                            VStack(alignment: .leading, spacing: Tokens.Spacing.xxs) {
                                Text(patient.fullName)
                                    .font(Tokens.Typography.subheadingRelative)
                                    .foregroundStyle(
                                        Tokens.Palette.textPrimary.resolve(for: scheme)
                                    )

                                Text("\(patient.mrn) · \(record.procedureName)")
                                    .font(Tokens.Typography.captionRelative)
                                    .foregroundStyle(
                                        Tokens.Palette.textSecondary.resolve(for: scheme)
                                    )
                            }

                            Spacer(minLength: Tokens.Spacing.sm)

                            Image(systemName: "chevron.right")
                                .font(Tokens.Typography.captionRelative)
                                .foregroundStyle(
                                    Tokens.Palette.textDisabled.resolve(for: scheme)
                                )
                                .accessibilityHidden(true)
                        }
                        .frame(minHeight: Tokens.minimumTouchTarget)
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .accessibilityElement(children: .combine)
                    .accessibilityAddTraits(.isButton)
                }

                LazyVGrid(
                    columns: [
                        GridItem(.flexible(), alignment: .topLeading),
                        GridItem(.flexible(), alignment: .topLeading),
                    ],
                    spacing: Tokens.Spacing.md
                ) {
                    FieldRow(
                        label: L10n.string("finance.net"),
                        value: record.netAmount.display(record.currency)
                    )
                    FieldRow(
                        label: L10n.string("finance.paid"),
                        value: record.paidAmount.display(record.currency)
                    )
                    FieldRow(
                        label: L10n.string("finance.balance"),
                        value: record.balance.display(record.currency),
                        tone: record.balance.isZero ? .success : .warning
                    )
                    FieldRow(
                        label: L10n.string("finance.agency"),
                        value: record.agencyName
                    )
                }

                HStack(spacing: Tokens.Spacing.sm) {
                    Badge(
                        record.paymentStatus.localizedName,
                        tone: FinanceScreen.statusTone(record.paymentStatus),
                        symbol: record.balance.isZero ? "checkmark.circle" : "hourglass"
                    )

                    Spacer(minLength: 0)
                }

                if !record.livePayments.isEmpty {
                    VStack(alignment: .leading, spacing: Tokens.Spacing.sm) {
                        Text(L10n.string("finance.payments.title"))
                            .font(Tokens.Typography.captionRelative)
                            .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))

                        ForEach(record.payments) { payment in
                            HStack(spacing: Tokens.Spacing.sm) {
                                Text(payment.amount.display(payment.currency))
                                    .font(Tokens.Typography.calloutRelative)
                                    .foregroundStyle(
                                        (payment.isReversed
                                            ? Tokens.Palette.textDisabled
                                            : Tokens.Palette.textPrimary).resolve(for: scheme)
                                    )
                                    .strikethrough(payment.isReversed)

                                Text(payment.method.localizedName)
                                    .font(Tokens.Typography.captionRelative)
                                    .foregroundStyle(
                                        Tokens.Palette.textSecondary.resolve(for: scheme)
                                    )

                                Spacer(minLength: Tokens.Spacing.sm)

                                if payment.isReversed {
                                    Badge(
                                        L10n.string("finance.payment.reversed"),
                                        tone: .neutral,
                                        symbol: "arrow.uturn.backward"
                                    )
                                } else {
                                    Button(L10n.string("finance.reverse")) { reversing = payment }
                                        .font(Tokens.Typography.captionRelative)
                                        .frame(minHeight: Tokens.minimumTouchTarget)
                                        .foregroundStyle(
                                            Tokens.Palette.critical.resolve(for: scheme)
                                        )
                                }
                            }
                            .accessibilityElement(children: .combine)
                        }
                    }
                }

                if !record.balance.isZero, record.cancelledAt == nil {
                    PrimaryButton(
                        title: L10n.string("finance.recordPayment"),
                        isBusy: state.busyId == record.id,
                        isEnabled: state.busyId == nil
                    ) {
                        paying = record
                    }
                }
            }
        }
    }

    // MARK: - Helpers

    static func statusTone(_ status: PaymentStatus) -> Tone {
        switch status {
        case .paid: return .success
        case .partial: return .warning
        case .pending: return .info
        case .refunded, .cancelled: return .neutral
        }
    }

    /// Older debt is redder. The label beside it says which bucket it is, so
    /// the colour is emphasis rather than the message.
    static func ageingTone(_ bucket: String) -> Tone {
        switch bucket {
        case "current": return .info
        case "d30": return .warning
        default: return .critical
        }
    }

    private func reload() async {
        await model.load()
        state = model.currentState()
    }
}

/// Taking a payment against one invoice.
struct PaymentSheet: View {
    @Environment(\.colorScheme) private var scheme
    @Environment(\.dismiss) private var dismiss

    let record: FinanceRecord
    let submit: (String, PaymentMethod, String?) async -> Void

    @State private var amount = ""
    @State private var method: PaymentMethod = .bankTransfer
    @State private var reference = ""
    @State private var busy = false

    var body: some View {
        FormScaffold(
            title: L10n.string("finance.recordPayment"),
            subtitle: record.patient?.fullName
        ) {
            VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                Card(tone: .info) {
                    FieldRow(
                        label: L10n.string("finance.balance"),
                        value: record.balance.display(record.currency)
                    )
                }

                LabelledField(
                    label: L10n.string("finance.amount"),
                    text: $amount,
                    isSecure: false,
                    contentType: .plain,
                    keyboard: .decimal
                )

                VStack(alignment: .leading, spacing: Tokens.Spacing.xs) {
                    Text(L10n.string("finance.method"))
                        .font(Tokens.Typography.captionRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))

                    Picker(L10n.string("finance.method"), selection: $method) {
                        ForEach(PaymentMethod.allCases, id: \.self) { option in
                            Text(option.localizedName).tag(option)
                        }
                    }
                    .pickerStyle(.menu)
                    .frame(minHeight: Tokens.minimumTouchTarget)
                }

                LabelledField(
                    label: L10n.string("finance.reference"),
                    text: $reference,
                    isSecure: false,
                    contentType: .plain,
                    keyboard: .default
                )

                PrimaryButton(
                    title: L10n.string("finance.recordPayment"),
                    isBusy: busy,
                    isEnabled: PaymentSheet.normalised(amount) != nil
                ) {
                    guard let normalised = PaymentSheet.normalised(amount) else { return }

                    busy = true
                    await submit(
                        normalised,
                        method,
                        reference.isEmpty ? nil : reference
                    )
                    busy = false
                }

                Button(L10n.string("common.cancel")) { dismiss() }
                    .frame(maxWidth: .infinity, minHeight: Tokens.minimumTouchTarget)
                    .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
            }
        }
    }

    /**
     * The amount as the server wants it: a decimal string with a full stop.
     *
     * A Turkish keyboard produces a comma, and sending "1.250,00" would be read
     * as one thousand two hundred and fifty *point* nothing — or refused. The
     * conversion is a normalisation, not arithmetic: the digits are unchanged.
     */
    static func normalised(_ input: String) -> String? {
        let trimmed = input
            .trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: ",", with: ".")

        guard !trimmed.isEmpty, let value = Decimal(string: trimmed), value > 0 else { return nil }

        return trimmed
    }
}


/**
 * Undoing a payment.
 *
 * A reason is required and the row is kept: a payment entered and undone is
 * part of what happened, and a ledger that forgets its corrections is one
 * nobody can audit. The server enforces both; this asks for the sentence while
 * whoever is undoing it still remembers why.
 */
struct ReversalSheet: View {
    @Environment(\.colorScheme) private var scheme
    @Environment(\.dismiss) private var dismiss

    let payment: PaymentEntry
    let submit: (String) async -> Void

    @State private var reason = ""
    @State private var busy = false

    var body: some View {
        FormScaffold(
            title: L10n.string("finance.reverse"),
            subtitle: payment.amount.display(payment.currency)
        ) {
            VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                LabelledField(
                    label: L10n.string("finance.reverseReason"),
                    text: $reason,
                    isSecure: false,
                    contentType: .plain,
                    keyboard: .default
                )

                PrimaryButton(
                    title: L10n.string("finance.reverse"),
                    isBusy: busy,
                    isEnabled: !reason.trimmingCharacters(in: .whitespaces).isEmpty && !busy
                ) {
                    busy = true
                    await submit(reason)
                    busy = false
                }

                Button(L10n.string("common.cancel")) { dismiss() }
                    .frame(maxWidth: .infinity, minHeight: Tokens.minimumTouchTarget)
                    .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
            }
        }
    }
}

File: android/feature/finance-ui/src/main/kotlin/xyz/klinik/feature/finance/ui/FinanceScreen.kt
package xyz.klinik.feature.finance.ui

import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import xyz.klinik.design.Tokens
import xyz.klinik.design.klinikColor
import xyz.klinik.feature.finance.FinancePhase
import xyz.klinik.feature.finance.FinanceState
import xyz.klinik.network.Currency
import xyz.klinik.network.ExchangeRate
import xyz.klinik.network.FinanceRecord
import xyz.klinik.network.PaymentStatus
import xyz.klinik.network.Totals
import xyz.klinik.network.UiText

/** Text the screen needs, resolved by the caller from string resources. */
data class FinanceStrings(
    val title: String,
    val notPermitted: String,
    val noRecords: String,
    val loadMore: String,
    val currency: String,
    val status: String,
    val allStatuses: String,
    val statusName: (PaymentStatus) -> String,
    val outstandingTitle: String,
    val collectionsTitle: String,
    val ratesTitle: String,
    val ratesNone: String,
    val recordsTitle: String,
    val ageingName: (String) -> String,
    val net: String,
    val paid: String,
    val balance: String,
    val recordPayment: String,
    val reverse: String,
    val recordCount: (Int) -> String,
    val totalsIncomplete: String,
    val message: (UiText) -> String,
)

/**
 * What has been billed and what has been paid (spec M11).
 *
 * The two reports are above the ledger because they are the two questions a
 * clinic actually asks — who owes us, and did the money arrive — and neither
 * is answerable by reading the list. The list is underneath for the one row
 * somebody came to find.
 *
 * Every figure on this screen is a string the server produced. Nothing here
 * adds up.
 */
@Composable
fun FinanceScreen(
    state: FinanceState,
    strings: FinanceStrings,
    onChooseCurrency: (Currency) -> Unit,
    onChooseStatus: (PaymentStatus?) -> Unit,
    onLoadMore: () -> Unit,
    onRecordPayment: (FinanceRecord) -> Unit,
    onReverse: (FinanceRecord) -> Unit,
    modifier: Modifier = Modifier,
) {
    Surface(color = klinikColor("background"), modifier = modifier.fillMaxSize()) {
        when (state.phase) {
            FinancePhase.Loading -> Centered { CircularProgressIndicator() }

            // Not an error and not a clinic that billed nothing: this account
            // may not see the screen, which reads differently.
            FinancePhase.NotPermitted -> Centered {
                Text(
                    strings.notPermitted,
                    color = klinikColor("textSecondary"),
                    textAlign = TextAlign.Center,
                )
            }

            FinancePhase.Loaded -> Column(
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(Tokens.Spacing.lg),
                verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.lg),
            ) {
                Text(
                    strings.title,
                    fontSize = Tokens.Typography.heading.size,
                    fontWeight = Tokens.Typography.heading.weight,
                    color = klinikColor("textPrimary"),
                    modifier = Modifier.semantics { heading() },
                )

                state.error?.let {
                    Text(strings.message(it), color = klinikColor("critical"))
                }

                Chips(
                    label = strings.currency,
                    options = Currency.entries,
                    chosen = state.currency,
                    name = { it.name },
                    onChoose = onChooseCurrency,
                )

                Chips(
                    label = strings.status,
                    // Null first: "everything" is the state the screen opens in.
                    options = listOf(null) + PaymentStatus.entries,
                    chosen = state.status,
                    name = { it?.let(strings.statusName) ?: strings.allStatuses },
                    onChoose = onChooseStatus,
                )

                state.outstanding?.let { report ->
                    Section(strings.outstandingTitle) {
                        Money(strings.balance, report.outstanding, strings)
                        Text(
                            strings.recordCount(report.recordCount),
                            fontSize = Tokens.Typography.caption.size,
                            color = klinikColor("textSecondary"),
                        )

                        report.ageing.forEach { bucket ->
                            Money(strings.ageingName(bucket.stringKey), bucket.totals, strings)
                        }
                    }
                }

                state.collections?.let { report ->
                    Section(strings.collectionsTitle) {
                        Money(strings.paid, report.received, strings)
                        Money(strings.net, report.net, strings)
                    }
                }

                Section(strings.ratesTitle) {
                    if (state.rates.isEmpty()) {
                        // Why a total is incomplete, said where the totals are
                        // rather than left for somebody to work out.
                        Text(strings.ratesNone, color = klinikColor("warning"))
                    } else {
                        state.rates.forEach { rate -> Rate(rate) }
                    }
                }

                Section(strings.recordsTitle) {
                    if (state.records.isEmpty()) {
                        Text(strings.noRecords, color = klinikColor("textSecondary"))
                    }

                    state.records.forEach { record ->
                        RecordRow(record, state, strings, onRecordPayment, onReverse)
                    }

                    if (state.hasMore) {
                        TextButton(
                            onClick = onLoadMore,
                            modifier = Modifier
                                .fillMaxWidth()
                                .heightIn(min = Tokens.minimumTouchTarget),
                        ) {
                            Text(strings.loadMore)
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun RecordRow(
    record: FinanceRecord,
    state: FinanceState,
    strings: FinanceStrings,
    onRecordPayment: (FinanceRecord) -> Unit,
    onReverse: (FinanceRecord) -> Unit,
) {
    val busy = state.busyId == record.id

    Column(
        modifier = Modifier.fillMaxWidth().padding(vertical = Tokens.Spacing.sm),
        verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs),
    ) {
        Row(modifier = Modifier.fillMaxWidth()) {
            Text(
                record.patient?.fullName ?: record.procedureName,
                color = klinikColor("textPrimary"),
                modifier = Modifier.weight(1f),
            )
            Text(
                strings.statusName(record.paymentStatus),
                fontSize = Tokens.Typography.caption.size,
                color = klinikColor("textSecondary"),
            )
        }

        Text(
            "${strings.net} ${record.currency.symbol}${record.netAmount} · " +
                "${strings.paid} ${record.currency.symbol}${record.paidAmount} · " +
                "${strings.balance} ${record.currency.symbol}${record.balance}",
            fontSize = Tokens.Typography.caption.size,
            // An overpaid record is not an ordinary one; it is money the clinic
            // owes back, and it should not read like a balance of the same kind.
            color = if (record.isOverpaid) klinikColor("warning") else klinikColor("textSecondary"),
        )

        Row(horizontalArrangement = Arrangement.spacedBy(Tokens.Spacing.sm)) {
            Button(
                onClick = { onRecordPayment(record) },
                enabled = !busy,
                modifier = Modifier.heightIn(min = Tokens.minimumTouchTarget),
            ) {
                Text(strings.recordPayment)
            }

            // Only where there is something to undo. A reversal needs a
            // payment, and offering it on a record with none is an offer that
            // can only fail.
            if (record.livePayments.isNotEmpty()) {
                TextButton(
                    onClick = { onReverse(record) },
                    enabled = !busy,
                    modifier = Modifier.heightIn(min = Tokens.minimumTouchTarget),
                ) {
                    Text(strings.reverse)
                }
            }
        }
    }
}

@Composable
private fun Rate(rate: ExchangeRate) {
    Row(modifier = Modifier.fillMaxWidth()) {
        Text(
            "${rate.base.name}/${rate.quote.name}",
            color = klinikColor("textSecondary"),
            modifier = Modifier.weight(1f),
        )
        Text("${rate.rate} · ${rate.validOn}", color = klinikColor("textPrimary"))
    }
}

@Composable
private fun Money(label: String, totals: Totals, strings: FinanceStrings) {
    Column(modifier = Modifier.fillMaxWidth()) {
        Row(modifier = Modifier.fillMaxWidth()) {
            Text(label, color = klinikColor("textSecondary"), modifier = Modifier.weight(1f))
            Text(
                "${totals.currency.symbol}${totals.converted}",
                color = klinikColor("textPrimary"),
            )
        }

        // The field a screen must not skip: a total that reads "128.400 ₺"
        // while two thousand pounds sit outside it has to say so, and then say
        // which.
        if (!totals.complete) {
            Text(
                strings.totalsIncomplete,
                fontSize = Tokens.Typography.caption.size,
                color = klinikColor("warning"),
            )

            totals.unconverted.forEach { amount ->
                Text(
                    "${amount.currency.symbol}${amount.amount}",
                    fontSize = Tokens.Typography.caption.size,
                    color = klinikColor("warning"),
                )
            }
        }
    }
}

@Composable
private fun <T> Chips(
    label: String,
    options: List<T>,
    chosen: T,
    name: (T) -> String,
    onChoose: (T) -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.xs)) {
        Text(
            label,
            fontSize = Tokens.Typography.caption.size,
            color = klinikColor("textSecondary"),
        )

        Row(
            modifier = Modifier.horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(Tokens.Spacing.sm),
        ) {
            options.forEach { option ->
                FilterChip(
                    selected = option == chosen,
                    onClick = { onChoose(option) },
                    label = { Text(name(option)) },
                    modifier = Modifier.heightIn(min = Tokens.minimumTouchTarget),
                )
            }
        }
    }
}

@Composable
private fun Section(title: String, content: @Composable ColumnScope.() -> Unit) {
    Surface(color = klinikColor("surface"), modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier.padding(Tokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.sm),
        ) {
            Text(
                title,
                fontSize = Tokens.Typography.subheading.size,
                fontWeight = Tokens.Typography.subheading.weight,
                color = klinikColor("textPrimary"),
                modifier = Modifier.semantics { heading() },
            )

            content()
        }
    }
}

@Composable
private fun Centered(content: @Composable () -> Unit) {
    Box(
        modifier = Modifier.fillMaxSize().padding(Tokens.Spacing.xl),
        contentAlignment = Alignment.Center,
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) { content() }
    }
}
